test/extensions/stats_sinks/hystrix/hystrix_test.cc

```diff
@@ -0,0 +1,348 @@
+  NiceMock<Server::MockInstance> server_;
+  Upstream::ClusterManager::ClusterInfoMap cluster_map_;
+
+  std::unique_ptr<HystrixSink> sink_;
+  NiceMock<Stats::MockSource> source_;
+  NiceMock<Upstream::MockClusterManager> cluster_manager_;
+};
+
+TEST_F(HystrixSinkTest, EmptyFlush) {
+  InSequence s;
+  Buffer::OwnedImpl buffer = createClusterAndCallbacks();
+  // Register callback to sink.
+  sink_->registerConnection(&callbacks_);
+  sink_->flush(source_);
+  std::string data_message = TestUtility::bufferToString(buffer);
+  validate_results(data_message, 0, 0, 0, 0, 0, window_size_);
+}
+
+TEST_F(HystrixSinkTest, BasicFlow) {
+  InSequence s;
+  Buffer::OwnedImpl buffer = createClusterAndCallbacks();
+  // Register callback to sink.
+  sink_->registerConnection(&callbacks_);
+
+  // Arbitrary numbers for testing. Make sure error > timeout.
+  uint64_t success_step = 7;
+  uint64_t error_step = 17;
+  uint64_t timeout_step = 3;
+  uint64_t rejected_step = 8;
+
+  for (uint64_t i = 0; i < (window_size_ + 1); i++) {
+    buffer.drain(buffer.length());
+    ON_CALL(cluster1_.error_5xx_counter_, value()).WillByDefault(Return((i + 1) * error_step));
+    ON_CALL(cluster1_.success_counter_, value()).WillByDefault(Return((i + 1) * success_step));
+    cluster1_.cluster_info_->stats().upstream_rq_timeout_.add(timeout_step);
+    cluster1_.cluster_info_->stats().upstream_rq_pending_overflow_.add(rejected_step);
+    sink_->flush(source_);
+  }
+
+  std::string rolling_map = sink_->printRollingWindows();
+  EXPECT_NE(std::string::npos, rolling_map.find(cluster1_name_ + ".total"));
+
+  // Check stream format and data.
+  validate_results(TestUtility::bufferToString(buffer), success_step, error_step, timeout_step, 0,
+                   rejected_step, window_size_);
+
+  // Check the values are reset.
+  buffer.drain(buffer.length());
+  sink_->resetRollingWindow();
+  sink_->flush(source_);
+  validate_results(TestUtility::bufferToString(buffer), 0, 0, 0, 0, 0, window_size_);
+}
+
+//
+TEST_F(HystrixSinkTest, Disconnect) {
+  InSequence s;
+  Buffer::OwnedImpl buffer = createClusterAndCallbacks();
+
+  sink_->flush(source_);
+  EXPECT_EQ(buffer.length(), 0);
+
+  // Register callback to sink.
+  sink_->registerConnection(&callbacks_);
+  sink_->flush(source_);
+  EXPECT_EQ(getStreamField(TestUtility::bufferToString(buffer), "rollingCountSuccess"), "0");
+  EXPECT_NE(buffer.length(), 0);
+
+  // Disconnect.
+  buffer.drain(buffer.length());
+  sink_->unregisterConnection(&callbacks_);
+  sink_->flush(source_);
+  EXPECT_EQ(buffer.length(), 0);
+
+  // Reconnect.
+  buffer.drain(buffer.length());
+  sink_->registerConnection(&callbacks_);
+  sink_->flush(source_);
+  EXPECT_EQ(getStreamField(TestUtility::bufferToString(buffer), "rollingCountSuccess"), "0");
+  EXPECT_NE(buffer.length(), 0);
+}
+
+TEST_F(HystrixSinkTest, AddAndRemoveClusters) {
+  InSequence s;
+
+  // Arbitrary numbers for testing. Make sure error > timeout.
+  uint64_t success_step = 7;
```

Comment:
nit: make these const.

I'm not sure I believe these are completely arbitrary though; I think some must be dependent on others.  Maybe comment those inline?

e.g.
```
  uint64_t error_step = 17;  // must be larger than xxxx + yyyyy +  ....
```
it seems unlikely that error > timeout is sufficient, right? If it just had to be > timeout I'd have expected you to set it to 2.

Reply:
Well, they are. Will set them to running values so it will be more obvious. And there aren't many restrictions on the actual values for the calculation purposes.